### PR TITLE
Request monitoring

### PR DESCRIPTION
**Pull request by krzysztofryk**
Implement monitoring of the size of responses for the given requests

## Description
Implement an additional Collector and Comparator that will allow one to sum the transferred data of the responses for the given/filtered requests. Added a simple report that shows the information about the overall amount of the transferred data and the table of ordered requests (desc) and their attributes (URL as a link, Size in kilobytes)

## Motivation and Context
Consider an example where one does not configure the images renditions behaviour properly and effectively downloads the original version of these on mobile devices. It is relatively easy to overlook such an issue in the ongoing development without automating this test case.
[#501](https://github.com/Cognifide/aet/issues/501)

## Screenshots (if appropriate)
n/a

## Upgrade notes (if appropriate)
n/a

## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [x] I have reviewed (and updated if needed) the documentation regarding this change

---
I hereby agree to the terms of the AET Contributor License Agreement.

### CODE REVIEW

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResultBuilder.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.cognifide.aet.job.common.comparators.requestmonitoring.utils;
+
+public class RequestMonitoringResultBuilder {
```
Comment (tenzo) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResultBuilder.java]: I consider this pattern is not needed in this context - correct me if I'm wrong but this builder is used only in one place and all of the fields seem to be mandatory. What is more - this is very simple object (2 fields) - I don't see a need to add this - constructor of `RequestMonitoringResult` should be sufficient.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResults.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * the License.
+ */
+package com.cognifide.aet.job.common.comparators.requestmonitoring.utils;
+
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+public class RequestMonitoringResults {
+
+  private final SortedSet<RequestMonitoringResult> results;
```
Comment (tenzo) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResults.java]: I would change it's type to `Collection` if there are no other requirements which force using this specific interface. I don't think explicitly informing this collection saves order is necessary in this context.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResult.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.cognifide.aet.job.common.comparators.requestmonitoring.utils;
+
+public class RequestMonitoringResult implements Comparable<RequestMonitoringResult> {
+
+  private final String url;
+
+  private final double size;
```
Comment (tenzo) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResult.java]: a comment with the unit (`//kB`) would be helpful there as it's included in other files :)

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/RequestMonitoringComparator.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      String maxSizeString = parameters.get(MAX_SIZE);
+      if (StringUtils.isNotEmpty(maxSizeString)) {
+        result = Long.valueOf(maxSizeString);
+      }
+    }
+    return result;
+  }
+
+  @Override
+  public void setParameters(Map<String, String> params) throws ParametersException {
```
[the remark this answers is not in the document]
Comment (krzysztofryk) [reply core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/RequestMonitoringComparator.java]: Done

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResult.java (hunk 3 above)
Comment (krzysztofryk) [reply core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResult.java]: Done

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResult.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    return url;
+  }
+
+  public double getSize() {
+    return size;
+  }
+
+  @Override
+  public int compareTo(RequestMonitoringResult o) {
+    return this.size > o.getSize() ? -1 : 1;
```
[the remark this answers is not in the document]
Comment (krzysztofryk) [reply core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResult.java]: Done.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResultBuilder.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+  private String url; //kB
```
[the remark this answers is not in the document]
Comment (krzysztofryk) [reply core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResultBuilder.java]: Done.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResults.java (hunk 2 above)
Comment (krzysztofryk) [reply core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResults.java]: Done.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResultBuilder.java (hunk 1 above)
Comment (krzysztofryk) [reply core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/utils/RequestMonitoringResultBuilder.java]: In principle I agree, I was a little bit reluctant on adding the logic of conversion from bytes to kB to the data object (RequestMonitoringResult). Actually, that was the only reason I decided to introduce this pattern.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/RequestMonitoringComparator.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+          .saveArtifactInJsonFormat(comparatorProperties, collectedResult);
+
+      if (maxSize != null) {
+        result = new ComparatorStepResult(artifactId,
+            collectedResult.getTotalSize() > maxSize ? ComparatorStepResult.Status.FAILED
+                : ComparatorStepResult.Status.PASSED);
+      } else {
+        result = new ComparatorStepResult(artifactId, ComparatorStepResult.Status.PASSED);
+      }
+
```
Comment (dzuma) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/requestmonitoring/RequestMonitoringComparator.java]: Empty lines

**Comment by annabudziszewska** [comment]
Two bugs reported: #599, #600

**Pull request merged**